### PR TITLE
Allow to provide custom levels and custom colors

### PR DESCRIPTION
**Pull request by matteozambon89**
This should provide the possibility to handle custom levels and custom colors, especially through CLI.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "pino-pretty",
-  "version": "7.1.0",
+  "version": "7.2.0",
```
Comment (mcollina) [inline package.json:3]: We will bump this ourselves, please revert.

File: package.json [hunk 2]
```diff
@@ -28,8 +28,7 @@
   },
   "homepage": "https://github.com/pinojs/pino-pretty#readme",
   "precommit": [
-    "lint",
-    "test"
+    "lint"
```
Comment (mcollina) [inline package.json:31]: please revert this change

File: lib/colors.js [hunk 3]
```diff
@@ -16,9 +16,10 @@ const plain = {
 }
 
 const { createColors } = require('colorette')
-const { white, bgRed, red, yellow, green, blue, gray, cyan } = createColors({ useColor: true })
+const availableColors = createColors({ useColor: true })
+const { white, bgRed, red, yellow, green, blue, gray, cyan } = availableColors
 
-const colored = {
+let colored = {
```
Comment (mcollina) [inline lib/colors.js:22]: Unfortunately this change is not ok. We should not be changing a singleton value in this way.

File: package.json (hunk 1 above)
Comment (matteozambon89) [reply package.json:3]: That's great! I'll revert this

File: package.json (hunk 2 above)
Comment (matteozambon89) [reply package.json:31]: I will revert as soon as I fix the tests, currently the coverage is under 100% and it's failing to commit.
I wouldn't merge this PR until the tests are with 100% coverage anyway

File: lib/colors.js (hunk 3 above)
Comment (matteozambon89) [reply lib/colors.js:22]: You are totally right, I will adapt to avoid the singleton mutation.

File: bin.js [hunk 4]
```diff
@@ -36,6 +36,8 @@ args
   .option(['e', 'errorProps'], 'Comma separated list of properties on error objects to show (`*` for all properties) (defaults to ``)')
   .option(['l', 'levelFirst'], 'Display the log level as the first output field')
   .option(['L', 'minimumLevel'], 'Hide messages below the specified log level')
+  .option(['X', 'customLevels'], 'Override default levels (`-cl err:99,info:1`)')
+  .option(['Y', 'customColors'], 'Override default colors using names from https://www.npmjs.com/package/colorette (`-cc err:red,info:blue`)')
```
Comment (matteozambon89) [inline bin.js:40]: @mcollina `-X` and `-Y` are just random 1 char option names... do you have any suggestions to how we should call these options?

I've tried to remove the 1 char name, but it defaults the name to `-C` on both.

Comment (mcollina) [reply bin.js:40]: Let's just avoid/remove the one-char option.

Comment (matteozambon89) [reply bin.js:40]: I'll try that again and post here the help result 👍

Comment (matteozambon89) [reply bin.js:40]: Removing the one-char option is not a solution, `customColors` and `customLevels` now have a default `-C` which is also used for `--config`. See example below:
```sh
  Usage: pino pretty [options] [command]
  
  Commands:
    help     Display help
    version  Display version
  
  Options:
    -c, --colorize              Force adding color sequences to the output
    -C, --config                specify a path to a json file containing the pino-pretty options
    -f, --crlf                  Append CRLF instead of LF to formatted lines
    -C, --customColors          Override default colors using names from https://www.npmjs.com/package/colorette (`-cc err:red,info:blue`)
    -C, --customLevels          Override default levels (`-cl err:99,info:1`)
    -k, --errorLikeObjectKeys   Define which keys contain error objects (`-k err,error`) (defaults to `err,error`)
    -e, --errorProps            Comma separated list of properties on error objects to show (`*` for all properties) (defaults to ``)
    -h, --help                  Output usage information
    -H, --hideObject            Hide objects from output (but not error object)
    -i, --ignore                Ignore one or several keys: (`-i time,hostname`)
    -l, --levelFirst            Display the log level as the first output field
    -L, --levelKey [value]      Detect the log level under the specified key (defaults to "level")
    -b, --levelLabel [value]    Output the log level using the specified label (defaults to "levelLabel")
    -o, --messageFormat         Format output of message
    -m, --messageKey [value]    Highlight the message under the specified key (defaults to "msg")
    -L, --minimumLevel          Hide messages below the specified log level
    -S, --singleLine            Print all non-error objects on a single line
    -a, --timestampKey [value]  Display the timestamp from the specified key (defaults to "time")
    -t, --translateTime         Display epoch timestamps as UTC ISO format or according to an optional format string (default ISO 8601)
    -v, --version               Output the version number
  
  Examples:
...
```

Any chance we can come up with a one-char for these two options?

Comment (matteozambon89) [reply bin.js:40]: See https://github.com/leo/args/issues/145

**Review by mcollina — APPROVED** [review]
lgtm

**Review by mcollina — CHANGES_REQUESTED** [review]
tests must be re-enabled

**Comment by matteozambon89** [comment]
@mcollina all done! Let me know if you can merge and deploy it please

File: lib/colors.js [hunk 5]
```diff
@@ -30,28 +31,57 @@ const colored = {
   greyMessage: gray
 }
 
-function colorizeLevel (level, colorizer) {
+function resolveCustomColored (customColors) {
```
Comment (jsumners) [inline lib/colors.js:34]: Is this supposed to be?:

```suggestion
function resolveCustomColors (customColors) {
```

File: lib/colors.js [hunk 6]
```diff
@@ -66,6 +96,12 @@ coloredColorizer.greyMessage = colored.greyMessage
  * colors as the integer `level` and will also default to `USERLVL` if the given
  * string is not a recognized level name.
  */
-module.exports = function getColorizer (useColors = false) {
-  return useColors ? coloredColorizer : plainColorizer
+module.exports = function getColorizer (useColors = false, customColors) {
```
Comment (jsumners) [inline lib/colors.js:99]: `customColors` is not documented in the docblock.

File: lib/utils.js [hunk 7]
```diff
@@ -223,9 +223,9 @@ function prettifyErrorLog ({
  * `undefined` will be returned. Otherwise, a string from the specified
  * `colorizer` is returned.
  */
-function prettifyLevel ({ log, colorizer = defaultColorizer, levelKey = LEVEL_KEY }) {
+function prettifyLevel ({ log, colorizer = defaultColorizer, levelKey = LEVEL_KEY, customLevels, customLevelNames }) {
```
Comment (jsumners) [inline lib/utils.js:226]: The new parameters are not documented in the docblock.

File: lib/utils.js [hunk 8]
```diff
@@ -245,12 +245,12 @@ function prettifyLevel ({ log, colorizer = defaultColorizer, levelKey = LEVEL_KE
  * key is not a string, then `undefined` will be returned. Otherwise, a string
  * that is the prettified message.
  */
-function prettifyMessage ({ log, messageFormat, messageKey = MESSAGE_KEY, colorizer = defaultColorizer, levelLabel = LEVEL_LABEL }) {
+function prettifyMessage ({ log, messageFormat, messageKey = MESSAGE_KEY, colorizer = defaultColorizer, levelLabel = LEVEL_LABEL, customLevels, levelKey = LEVEL_KEY }) {
```
Comment (jsumners) [inline lib/utils.js:248]: The new parameters are not documented in the docblock.

**Review by jsumners — CHANGES_REQUESTED** [review]
Please update documentation, including the Readme.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by matteozambon89** [comment]
@mcollina and @jsumners can this be merged to `master`?

File: package.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "pino-pretty",
-  "version": "7.2.0",
+  "version": "7.1.0",
```
Comment (jsumners) [inline package.json:3]: Please don't change this. We will handle it during release.

**Review by jsumners — CHANGES_REQUESTED** [review]
Can you perform a rebase?

```sh
$ git remote add upstream https://github.com/pinojs/pino-pretty.git
$ git fetch upstream master
$ git rebase upstream/master
```

**Comment by matteozambon89** [comment]
@jsumners I've removed `package.json` from my commits as the changes shouldn't have been committed in the first place.

**Review by jsumners — APPROVED** [review]
Looks good to me.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
can you check CI? It was failing

**Comment by dotcore** [comment]
I am also very interested in this feature, so I looked into why the CI is failing. For some probably non-intentional reason in commit id c94f89c in file `test/basic.test.js`, following code gets changed to: 

```
    const expectedLines = [
      '    msg: {',
      '      "b": {',
      '        "c": "d"',
      '      },',
      '      "a": "[Circular ~]"',
      '    }'
    ]
```
instead of:

```
    const expectedLines = [
      '    msg: {',
      '      "a": "[Circular]",',
      '      "b": {',
      '        "c": "d"',
      '      }',
      '    }'
    ]
```

After that the tests are all green.
I think this also could use a rebase to upstream again. I could also submit a pull request, if you want.

**Comment by matteozambon89** [comment]
> I am also very interested in this feature, so I looked into why the CI is failing. For some probably non-intentional reason in commit id [c94f89c](https://github.com/pinojs/pino-pretty/commit/c94f89ce3e4ab36387e6061118ebc6bdd98f2472) in file `test/basic.test.js`, following code gets changed to:
> 
> ```
>     const expectedLines = [
>       '    msg: {',
>       '      "b": {',
>       '        "c": "d"',
>       '      },',
>       '      "a": "[Circular ~]"',
>       '    }'
>     ]
> ```
> 
> instead of:
> 
> ```
>     const expectedLines = [
>       '    msg: {',
>       '      "a": "[Circular]",',
>       '      "b": {',
>       '        "c": "d"',
>       '      }',
>       '    }'
>     ]
> ```
> 
> After that the tests are all green. I think this also could use a rebase to upstream again. I could also submit a pull request, if you want.

Sorry if I've been AWOL, between Xmas holidays and New Year "back to work" I haven't had time to look into this issue.

Thanks @dotcore for your input. The commit was made somewhat intentionally as the local test was failing and wasn't allowing me to push. I'll revert that commit and push again to see if it fixes the issue.

I will also do a rebase to upstream, so we can hopefully close this PR.

**Comment by matteozambon89** [comment]
@mcollina @jsumners @dotcore reverted the test changes and rebased to upstream

**Comment by dotcore** [comment]
> @mcollina @jsumners @dotcore reverted the test changes and rebased to upstream

One typo is left.

`'      "a": "[Circular]",',`

instead of

`'      "a": "[Circular],"',`

**Comment by matteozambon89** [comment]
@dotcore I've updated that, although my tests keep failing for the test/basic.test.js as the format is different on my machine.

I'm on MacOS with node 16, which environment are you running the tests from?

**Comment by dotcore** [comment]
> @dotcore I've updated that, although my tests keep failing for the test/basic.test.js as the format is different on my machine.
> 
> I'm on MacOS with node 16, which environment are you running the tests from?

I'm on MacOS with node 17.4.0, but I also tested on node 16.13.2 and 12.22.9 (Switched with nvm).

**Comment by matteozambon89** [comment]
@dotcore for some reason this is what I see when I run `npm run test`, can you confirm this doesn't happen to you and all tests are passing?

```
> pino-pretty@7.3.0 test
> tap --100 --color

 FAIL  test/basic.test.js
 ✖ should be equal

  test/basic.test.js                                         
  701 |         lines.shift(); lines.pop()                   
  702 |         for (let i = 0; i < lines.length; i += 1) {  
> 703 |           t.equal(lines[i], expectedLines[i])        
      | ------------^                                        
  704 |         }                                            
  705 |         cb()                                         
  706 |       }                                              

  --- expected              
  +++ actual                
  @@ -1,1 +1,1 @@           
  -      "a": "[Circular]", 
  +      "b": {             

  test: test/basic.test.js basic prettifier tests prettifies msg object with
    circular references
  stack: |
    Writable.write [as _write] (test/basic.test.js:703:13)
    doWrite (node_modules/readable-stream/lib/_stream_writable.js:409:139)
    writeOrBuffer (node_modules/readable-stream/lib/_stream_writable.js:398:5)
    Writable.write (node_modules/readable-stream/lib/_stream_writable.js:307:11)
    Pino.write (node_modules/pino/lib/proto.js:190:10)
    Pino.LOG [as info] (node_modules/pino/lib/tools.js:58:21)
    Test.<anonymous> (test/basic.test.js:711:9)
    Test.<anonymous> (test/basic.test.js:685:5)
    Object.<anonymous> (test/basic.test.js:34:1)
    Module.replacementCompile (node_modules/append-transform/index.js:60:13)
    Object.<anonymous> (node_modules/append-transform/index.js:64:4)

 FAIL  test/basic.test.js
 ✖ should be equal

  test/basic.test.js                                         
  701 |         lines.shift(); lines.pop()                   
  702 |         for (let i = 0; i < lines.length; i += 1) {  
> 703 |           t.equal(lines[i], expectedLines[i])        
      | ------------^                                        
  704 |         }                                            
  705 |         cb()                                         
  706 |       }                                              

  --- expected      
  +++ actual        
  @@ -1,1 +1,1 @@   
  -      "b": {     
  +        "c": "d" 

  test: test/basic.test.js basic prettifier tests prettifies msg object with
    circular references
  stack: |
    Writable.write [as _write] (test/basic.test.js:703:13)
    doWrite (node_modules/readable-stream/lib/_stream_writable.js:409:139)
    writeOrBuffer (node_modules/readable-stream/lib/_stream_writable.js:398:5)
    Writable.write (node_modules/readable-stream/lib/_stream_writable.js:307:11)
    Pino.write (node_modules/pino/lib/proto.js:190:10)
    Pino.LOG [as info] (node_modules/pino/lib/tools.js:58:21)
    Test.<anonymous> (test/basic.test.js:711:9)
    Test.<anonymous> (test/basic.test.js:685:5)
    Object.<anonymous> (test/basic.test.js:34:1)
    Module.replacementCompile (node_modules/append-transform/index.js:60:13)
    Object.<anonymous> (node_modules/append-transform/index.js:64:4)

 FAIL  test/basic.test.js
 ✖ should be equal

  test/basic.test.js                                         
  701 |         lines.shift(); lines.pop()                   
  702 |         for (let i = 0; i < lines.length; i += 1) {  
> 703 |           t.equal(lines[i], expectedLines[i])        
      | ------------^                                        
  704 |         }                                            
  705 |         cb()                                         
  706 |       }                                              

  --- expected      
  +++ actual        
  @@ -1,1 +1,1 @@   
  -        "c": "d" 
  +      },         

  test: test/basic.test.js basic prettifier tests prettifies msg object with
    circular references
  stack: |
    Writable.write [as _write] (test/basic.test.js:703:13)
    doWrite (node_modules/readable-stream/lib/_stream_writable.js:409:139)
    writeOrBuffer (node_modules/readable-stream/lib/_stream_writable.js:398:5)
    Writable.write (node_modules/readable-stream/lib/_stream_writable.js:307:11)
    Pino.write (node_modules/pino/lib/proto.js:190:10)
    Pino.LOG [as info] (node_modules/pino/lib/tools.js:58:21)
    Test.<anonymous> (test/basic.test.js:711:9)
    Test.<anonymous> (test/basic.test.js:685:5)
    Object.<anonymous> (test/basic.test.js:34:1)
    Module.replacementCompile (node_modules/append-transform/index.js:60:13)
    Object.<anonymous> (node_modules/append-transform/index.js:64:4)

 FAIL  test/basic.test.js
 ✖ should be equal

  test/basic.test.js                                         
  701 |         lines.shift(); lines.pop()                   
  702 |         for (let i = 0; i < lines.length; i += 1) {  
> 703 |           t.equal(lines[i], expectedLines[i])        
      | ------------^                                        
  704 |         }                                            
  705 |         cb()                                         
  706 |       }                                              

  --- expected               
  +++ actual                 
  @@ -1,1 +1,1 @@            
  -      }                   
  +      "a": "[Circular ~]" 

  test: test/basic.test.js basic prettifier tests prettifies msg object with
    circular references
  stack: |
    Writable.write [as _write] (test/basic.test.js:703:13)
    doWrite (node_modules/readable-stream/lib/_stream_writable.js:409:139)
    writeOrBuffer (node_modules/readable-stream/lib/_stream_writable.js:398:5)
    Writable.write (node_modules/readable-stream/lib/_stream_writable.js:307:11)
    Pino.write (node_modules/pino/lib/proto.js:190:10)
    Pino.LOG [as info] (node_modules/pino/lib/tools.js:58:21)
    Test.<anonymous> (test/basic.test.js:711:9)
    Test.<anonymous> (test/basic.test.js:685:5)
    Object.<anonymous> (test/basic.test.js:34:1)
    Module.replacementCompile (node_modules/append-transform/index.js:60:13)
    Object.<anonymous> (node_modules/append-transform/index.js:64:4)

 FAIL  test/basic.test.js
 ✖ should be equal

  test/basic.test.js                                                             
  931 |     Object.assign(fs, { close, writeSync })                              
  932 |     t.match(chunks.join(''), /INFO .+: this message has been buffered/)  
> 933 |     t.equal(closeCalled, false)                                          
      | ------^                                                                  
  934 |   })                                                                     
  935 |                                                                          
  936 |   t.test('stream usage', async (t) => {                                  

  test: test/basic.test.js basic prettifier tests does not call fs.close on stdout
    stream
  found: true
  wanted: false
  compare: ===
  stack: |
    Test.<anonymous> (test/basic.test.js:933:7)
    Test.<anonymous> (test/basic.test.js:910:5)
    Object.<anonymous> (test/basic.test.js:34:1)
    Module.replacementCompile (node_modules/append-transform/index.js:60:13)
    Object.<anonymous> (node_modules/append-transform/index.js:64:4)
```

**Comment by dotcore** [comment]
> @dotcore for some reason this is what I see when I run `npm run test`, can you confirm this doesn't happen to you and all tests are passing?

I can confirm that it works for me, also CI didn't fail. No idea why it does not work for you.

**Comment by matteozambon89** [comment]
> > @dotcore for some reason this is what I see when I run `npm run test`, can you confirm this doesn't happen to you and all tests are passing?
> 
> I can confirm that it works for me, also CI didn't fail. No idea why it does not work for you.

Who knows 😂

**Comment by dotcore** [comment]
> > > @dotcore for some reason this is what I see when I run `npm run test`, can you confirm this doesn't happen to you and all tests are passing?
> > 
> > 
> > I can confirm that it works for me, also CI didn't fail. No idea why it does not work for you.
> 
> Who knows 😂

Not sure if this may help, but maybe delete your node_modules and `npm i`, in case you have some outdated dependencies.

**Comment by matteozambon89** [comment]
@dotcore nope, that didn't solve the issue. Something between node and MacOS must be configured differently in this machine.

@mcollina can we have this PR merged to master?

**Comment by mcollina** [comment]
Done
